src/reducers/viewconfig.ts

```diff
@@ -128,140 +176,166 @@ const getAllDescendantIds = (state: ViewConfigReducer, nodeId: number) =>
         []
     );
 
-const deleteMany = (state: ViewConfigReducer, ids: number[]): ViewConfigReducer => {
-    state = [...state];
-    for (const id of ids) {
-        const index = state.findIndex(item => item.id === id);
-        state.splice(index, 1);
+const deleteAllReferences = (state: ViewConfigReducer, nodeId: number): ViewConfigReducer =>
+    state.map(stateNode => {
+        for (const [idx, childId] of stateNode.childIds.entries()) {
+            if (childId === nodeId) {
+                stateNode.childIds.splice(idx, 1);
+            }
+        }
+        return stateNode;
+    });
+
+const stateWithoutNodeReferences = (state: ViewConfigReducer, nodeId: number): ViewConfigReducer => {
+    const newState = [...state];
+
+    for (const id of [nodeId, ...getAllDescendantIds(state, nodeId)]) {
+        const index = getNodeIndex(id, state);
+
+        if (index) {
+            newState.splice(index, 1);
+        }
     }
-    return state;
-};
 
-export const denormalizeComponent = (item: NormalizedComponentConfig): ComponentConfig => {
-    item = { ...item };
-    delete item.id;
-    delete item.childIds;
-    delete item.name;
-    return item;
+    return deleteAllReferences(newState, nodeId);
 };
 
 // reducers
-
-const childIds = (state: number[], action: Action) => {
+const nodesToAdd = (state: ViewConfigReducer, action: Action): NormalizedComponentConfig[] => {
     switch (action.type) {
-        case 'ADD_VIEW_CONFIG_CHILD':
-            return [...state, action.payload.childId];
-        case 'DELETE_VIEW_CONFIG_CHILD':
-            return state.filter(id => id !== action.payload.childId);
-        case 'SORT_VIEW_CONFIG_CHILD':
-            return arrayMove(state, action.payload.oldIndex, action.payload.newIndex);
+        case 'ADD_VIEW_CONFIG_NODE':
+            return normalizeTree([action.payload.component], action.payload.collectionId, lastId(state));
+        case 'SWITCH_VIEW_CONFIG_NODE': {
+            const children = normalizeTree(
+                action.payload.component.subComponents || [],
+                action.payload.collectionId,
+                lastId(state)
+            );
+            children.pop();
+            return children;
+        }
         default:
-            return state;
+            return [];
     }
 };
 
 const node = (
-    state: NormalizedComponentConfig | null,
+    state: NormalizedComponentConfig | ComponentConfig | null,
```

Comment:
state gets cast to `NormalizedComponentConfig` everywhere in this function, and it also doesn't get called with null anymore.
So state can just be `NormalizedComponentConfig`, right?